plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java

```diff
@@ -410,6 +411,35 @@ private static <K, V> V getWithValueHolder(Cache<K, AtomicReference<V>> cache, K
         return value;
     }
 
+    private static <K, V> V getWithValueHolder(
+            Cache<K, AtomicReference<V>> cache, K key,
+            Predicate<V> shouldReload,
+            Supplier<V> loader,
+            Function<V, V> reloader,
+            BiFunction<V, V, V> merger)
+    {
+        AtomicReference<V> valueHolder = uncheckedCacheGet(cache, key, AtomicReference::new);
+        V oldValue = valueHolder.get();
+        if (oldValue != null && !shouldReload.test(oldValue)) {
+            return oldValue;
+        }
+
+        V newValue = oldValue == null ? loader.get() : reloader.apply(oldValue);
+
+        if (newValue == null) {
+            throw new InvalidCacheLoadException("Failed to return a value for " + key);
+        }
+        V finalValue;
+        if (oldValue == null) {
+            finalValue = newValue;
+        }
+        else {
+            finalValue = merger.apply(oldValue, newValue);
+        }
+        valueHolder.compareAndSet(oldValue, finalValue);
```

Comment:
it looks like merger is used only when incremental loader is used, so could be part of same lambda. fewer parameters => better

however, concurrent loads can incrementally load (add) different columns, and we could use `accumulateAndGet` instead of `compareAndSet` to avoid overwriting concurrent loads.

Reply:
> Do you mean to do the `if oldValue == null` check in the lambda ?

no, i meant to have two lambdas: loader (when there was no old value) and incrementalLoader (when it was present)

anyway, will re-review again, no need to change anything here now